ext/bcmath/libbcmath/src/div.c

```diff
@@ -31,219 +31,446 @@
+		numerator_vectors[numerator_top_index - i - 1] += numerator_vectors[numerator_top_index - i] * BC_VECTOR_BOUNDARY_NUM;
+		numerator_vectors[numerator_top_index - i] = 0;
+	}
+	/* last */
+	quot_vectors[0] = numerator_vectors[0] / divisor_vector;
+}
 
-	if (digit == 0) {
-		memset(result, 0, size);
-	} else {
-		if (digit == 1) {
-			memcpy(result, num, size);
-		} else {
-			/* Initialize */
-			nptr = (unsigned char *) (num + size - 1);
-			rptr = (unsigned char *) (result + size - 1);
-			carry = 0;
-
-			while (size-- > 0) {
-				value = *nptr-- * digit + carry;
-				*rptr-- = value % BASE;
-				carry = value / BASE;
-			}
+/*
+ * Used when the divisor is split into 2 or more chunks.
+ * This use the restoring division algorithm.
+ * https://en.wikipedia.org/wiki/Division_algorithm#Restoring_division
+ */
+static inline void bc_standard_div(
+	BC_VECTOR *numerator_vectors, size_t numerator_arr_size,
+	BC_VECTOR *divisor_vectors, size_t divisor_arr_size, size_t divisor_len,
+	BC_VECTOR *quot_vectors, size_t quot_arr_size
+) {
+	size_t numerator_top_index = numerator_arr_size - 1;
+	size_t divisor_top_index = divisor_arr_size - 1;
+	size_t quot_top_index = quot_arr_size - 1;
+
+	BC_VECTOR div_carry = 0;
+
+	/*
+	 * Errors might occur between the true quotient and the temporary quotient calculated using only the high order digits.
+	 * For example, the true quotient of 240 / 121 is 1.
+	 * However, if it is calculated using only the high-order digits (24 / 12), we would get 2.
+	 * We refer to this value of 2 as the temporary quotient.
+	 * We also define E to be error between the true quotient and the temporary quotient,
+	 * which in this case, is 1.
+	 *
+	 * Another example: Calculating 999_0000_0000 / 1000_9999 with base 10000.
+	 * The true quotient is 9980, but if it is calculated using only the high-order digits (999 / 1000), we would get 0
+	 * If the temporary quotient is 0, we need to carry the digits to the next division, which is 999_0000 / 1000.
+	 * The new temporary quotient we get is 9990, with error E = 10.
+	 *
+	 * If calculate the temporary quotient using only one array of numerator and divisor, the error E can be larger than 1.
+	 * In other words, in the restoring division, the count of additions for restore increases significantly.
+	 *
+	 * Therefore, in order to keep the error within 1 and to limit the number of additions required for restoration to
+	 * at most one, adjust the number of high-order digits used to calculate the temporary quotient as follows.
+	 * - Adjust the number of digits of divisor used in the calculation to BC_VECTOR_SIZE + 1 digit. The missing digits are
+	 *   filled in from the next array element.
+	 * - Add digits to numerator in the same way as the number of digits adjusted by divisor.
+	 *
+	 * e.g.
+	 * numerator = 123456780000
+	 * divisor = 123456789
+	 * base = 10000
+	 * numerator_arr = [0, 5678, 1234]
+	 * divisor_arr = [6789, 2345, 1]
```

Comment:
Why is the split for the divisor favouring the low digits whereas the numerator favours the high digits?

Reply:
> the numerator favours the high digits

No, this isn't actually the case.  Written in an easy-to-understand manner, it looks like this:
```
numerator_arr = [0000, 5678, 1234]
```